api/proto/objects/revision.proto

```diff
@@ -0,0 +1,70 @@
+syntax = "proto3";
+
+package jj_api.objects;
+
+// A RevisionMask controls how much of a revision is returned.
+// For example, `jj log` will likely not care about files, while `jj status`
+// will want to know what files changed in a revision, and `jj diff` will want
+// to know both the content of the file and the content of the file in the
+// parent revision.
+message RevisionMask {
+  // Which files will be returned.
+  enum FilePathMask {
+    NONE = 0;
+    MODIFIED_FILES = 1;
+    ALL_FILES = 3;
+    // Include all files that were matched by the RevisionMask containing this
+    // RevisionMask. For example, if we had:
+    // RevisionMask(
+    //   files_to_include = MODIFIED_FILES,
+    //   files = FileMask(content = True),
+    //   parents = RevisionMask(
+    //     files_to_include = PARENT_FILES
+    //     files = FileMask(content = True),
+    //   ),
+    // )
+    // And we returned the revision @, which modfied the file foo, then both
+    // r.files["foo"].content and r.parent.files["foo"].content would be filled.
+    PARENT_FILES = 4;
+  };
+  FilePathMask files_to_include = 1;
+  repeated string additional_files = 2;
+  
+  bool rendered = 3;
+
+  bool file_content = 4;
+  bool file_hash = 5;
+  bool file_metadata = 6;
+
+  // How much of the parent revision to fill in.
+  RevisionMask parents = 7;
+}
+
+message File {
+  // If you query for MODIFIED_FILES, you may get a file that's been deleted.
+  bool exists = 1;
+  string hash = 2;
+  bytes content = 3;
+  // TODO: file metadata (eg. permissions)?
+}
+
+
+// By default, 
+message Revision {
+  string change_id = 1;
+  string commit_id = 2;
+  string description = 3;
+  bool empty = 4;
+  bool conflicts = 5;
+  bool mutable = 6;
+  
+  // The rendered template.
+  // Open question: Is a template a first-class citizen of jj, or is it specific
+  // to jj-cli?
+  string rendered = 7;
```

Comment:
Pulling @martinvonz's comment into here to discuss further in a single thread.

> My gut feeling is that templating should not be in the API.

Seems the two of you are disagreeing here, so I'll list what I think the consequences are of each option:
* If a template is a first-class citizen of jj, then you should be able to call `ListRevision(template="...")` and have it return `Revision(rendered = "...")`
* This means that any extension should be able to use your templates / template aliases that you defined
  * This means that the jj daemon will need to read your config to be aware of your templates
    * I personally don't have a problem with this - I think that the daemon is definitely going to need to read your config regardless, since I think it'll be required for `Revset`
* If a template is not a first-class citizen of jj, then it must be done purely from the jj cli.
  * Potential performance overhead of this, since jj-cli will need to *always* fill all fields just in case the template references them (unless you can do a significant amount of pre-processing to determine what properties of the template it's gonna reference).
  * Templates will be useable *only* for jj-cli (eg. extensions cannot use them).

After thinking about it in more detail, I think that adding it as a first-class citizen is probably my preferred choice. I can imagine that, for example, an extension such as `jj piper upload` might want to add a template alias so that a user could override it from printing:
```
Exported CLs:
  http://cl/643830362 qtoyoqum d5973671 cl/643830362 | foo
```

And instead print:
```
Exported CLs:
  http://cl/643830362 foo
```

Reply:
> I can imagine that, for example, an extension such as `jj piper upload` might want to add a template alias so that a user could override it from printing:

I thought the conclusion from the design doc was that the RPC is basically for adding new UIs while the Rust API is usually required for adding new commands. Did I misunderstand/misremember?

Reply:
This sounds like some miscommunication went on. I suspect that my `jj piper upload` example may have been a confusing choice of command, since it already exists, and wasn't implemented as an extension (since we don't support extensions, and even if we did, we'd require modifying `jj` for the extra backend support).

My assumption was that:
* Commands could be added via an extension (#3575)
* An extension would be a binary external to `jj` (and thus not neccesarily written in rust) that does whatever it wants under the hood via the jj API.

For example, if we chose to implement gerrit support as an extension rather than natively in jj, I assumed that it might look like (very pseudocode):

```python
#!/usr/bin/env python3

import jj

args = argparse.ArgumentParser(...).parse_args()
client = jj.client()
transaction = client.StartTransaction()
rev = client.GetRevision(jj.RevisionRequest(transaction=transaction, revset=args.revset))
rev = client.UpdateDescription(transaction=transaction, revision=rev, description=client.description + "Change-ID: ...")
client.EndTransaction(transaction)
subcommand.run(["git", "push", "--remote", "gerrit", f"{rev.commit_id}:refs/for/main"]
print(f"Sent, {rev} to gerrit for review")
```
My assumption In extension would be a binary external to `jj`, and that it wouldn't have to be implemented in rust. I assumed that I'd be able to write an extension in python, for example, and then use the RPCs to implement it.

Reply:
> This may have an perf overhead, and It also isn't clear on how to expose them correctly

My thoughts was that you would execute a `ListRevisions(revision_mask=RevisionMask(template="...")`, and then each `Revision` returned would have the template filled in.

If you executed `ListRevisions(revision_mask=RevisionMask(...))` (no template), then we wouldn't calculate the templates. This should have no performance overhead, since templates are only calculated when we explicitly request for them. In fact, this is likely actually cheaper, since if you calculated templates client side, you'd need to calculate *all* fields just in case they're needed for templates.

> My opinion on that is that we need general commit metadata anyways, so `piper upload` can set something like "key" => `codereview` "value" => `cl/NNNNN` which we then can standardize across forges and can make available in templates. This should require no extension work.

I think you may have misunderstood what my point was here. I was simply saying that an arbitrary extension may want access to templates to change how commits are printed. I was simply suggesting that the user might want to change the template from `seperate(" ", codereview_link(), change_id(), commit_id(), branches(), "|", description())` to `seperate(" ", codereview_link(), description())`, and that this wasn't easily possible if the extension doesn't have access to templates.